cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,544 @@
+    return circuit
+
+
+def _random_two_q_clifford(q_0, q_1, num_cfds, cfd_matrices, c1, s1, s1_x,
+                           s1_y):
+    idx_list = list(numpy.random.choice(11520, num_cfds))
+    circuit = circuits.Circuit()
+    for idx in idx_list:
+        circuit.append(
+            _two_qubit_clifford(q_0, q_1, idx, c1, s1, s1_x, s1_y))
+    inv_idx = _find_inv_matrix(protocols.unitary(circuit), cfd_matrices)
+    circuit.append(
+        _two_qubit_clifford(q_0, q_1, inv_idx, c1, s1, s1_x, s1_y))
+    return circuit
+
+
+def _find_inv_matrix(mat: numpy.ndarray, mat_sequence: numpy.ndarray) -> int:
+    mat_prod = numpy.einsum('ij,...jk->...ik', mat, mat_sequence)
+    diag_sums = list(numpy.absolute(numpy.einsum('...ii->...', mat_prod)))
+    idx = diag_sums.index(max(diag_sums))
+    return idx
+
+
+def _matrix_bar_plot(mat, z_label: str, kets=None, title=None) -> None:
+    num_rows, num_cols = mat.shape
+    indices = numpy.meshgrid(range(num_cols), range(num_rows))
+    x_indices = numpy.array(indices[1]).flatten()
+    y_indices = numpy.array(indices[0]).flatten()
+    z_indices = numpy.zeros(mat.size)
+
+    dx = numpy.ones(mat.size) * 0.3
+    dy = numpy.ones(mat.size) * 0.3
+
+    fig = pyplot.figure()
+    ax1 = fig.add_subplot(111, projection='3d')  # type: Axes3D
+
+    dz = mat.flatten()
+    ax1.bar3d(x_indices, y_indices, z_indices, dx, dy, dz, color=
+    '#ff0080', alpha=1.0)
+
+    ax1.set_zlabel(z_label)
+    ax1.set_zlim3d(min(0, numpy.amin(mat)), max(0, numpy.amax(mat)))
+
+    if kets is not None:
+        pyplot.xticks(numpy.arange(num_cols) + 0.15, kets)
+        pyplot.yticks(numpy.arange(num_rows) + 0.15, kets)
+
+    if title is not None:
+        ax1.set_title(title)
+
+
+def _plot_density_matrix(mat: numpy.ndarray) -> None:
+    a, _ = mat.shape
+    num_qubits = int(numpy.sqrt(a))
+    state_labels = [[0, 1]] * num_qubits
+    kets = []
+    for label in itertools.product(*state_labels):
+        kets.append('|' + str(list(label))[1:-1] + '>')
+    mat_re = numpy.real(mat)
+    mat_im = numpy.imag(mat)
+    _matrix_bar_plot(mat_re, r'Real($\rho$)', kets)
+    _matrix_bar_plot(mat_im, r'Imaginary($\rho$)', kets)
+
+
+def _gate_seq_to_mats(gate_seq: Sequence[ops.Gate]):
+    mat_rep = protocols.unitary(gate_seq[0])
+    for gate in gate_seq[1:]:
+        mat_rep = numpy.dot(protocols.unitary(gate), mat_rep)
+    return mat_rep
+
+
+def _two_qubit_clifford(q_0, q_1, idx, c1, s1, s1_x, s1_y):
+    idx_0 = int(idx / 480)
+    idx_1 = int((idx % 480) / 20)
+    idx_2 = idx - idx_0 * 480 - idx_1 * 20
+    yield _single_qubit_gates(c1[idx_0], q_0)
+    yield _single_qubit_gates(c1[idx_1], q_1)
+    if idx_2 == 1:
+        yield ops.CZ(q_0, q_1)
+        yield ops.Y(q_0) ** -0.5
+        yield ops.Y(q_1) ** 0.5
+        yield ops.CZ(q_0, q_1)
+        yield ops.Y(q_0) ** 0.5
+        yield ops.Y(q_1) ** -0.5
+        yield ops.CZ(q_0, q_1)
+        yield ops.Y(q_1) ** 0.5
+    elif 1 < idx_2 < 11:
```

Comment:
This needs comments. Where are the numbers from?

Also, make this 11 match the 10 in the next clause by using <= or >= in one of them.

Reply:
This function now has a docstring explaining how a two-qubit Clifford is generated.